DeleteUserDataActivity/backupAndDelete.ts

```diff
@@ -0,0 +1,472 @@
+import { FiscalCode } from "italia-ts-commons/lib/strings";
+import {
+  IResultIterator,
+  iteratorToArray
+} from "../utils/extensions/documentdb";
+import { MessageDeletableModel } from "../utils/extensions/models/message";
+import { MessageStatusDeletableModel } from "../utils/extensions/models/message_status";
+import { NotificationDeletableModel } from "../utils/extensions/models/notification";
+import { NotificationStatusDeletableModel } from "../utils/extensions/models/notification_status";
+import { ProfileDeletableModel } from "../utils/extensions/models/profile";
+import { DataFailure, IBlobServiceInfo } from "./types";
+import { saveDataToBlob } from "./utils";
+import { toDocumentDeleteFailure, toQueryFailure } from "./utils";
+
+import {
+  fromQueryEither,
+  QueryError
+} from "io-functions-commons/dist/src/utils/documentdb";
+
+/**
+ * Recursively consumes an iterator and executes operations on every item
+ * @param deleteSingle takes an item and delete it
+ * @param userDataBackup references about where to save data
+ * @param makeBackupBlobName takes an item and construct a name for the backup blob
+ * @param iterator an iterator of every result from the db
+ */
+const executeRecursiveBackupAndDelete = <T>(
+  deleteSingle: (item: T) => Promise<Either<QueryError, string>>,
+  userDataBackup: IBlobServiceInfo,
+  makeBackupBlobName: (item: T) => string,
+  iterator: IResultIterator<T>
+): TaskEither<
+  // tslint:disable-next-line: use-type-alias
+  DataFailure,
+  readonly T[]
+> => {
+  return (
+    tryCatch(iterator.executeNext, toError)
+      // this is just type lifting
+      .foldTaskEither<DataFailure, Option<readonly T[]>>(
+        e => fromLeft(toQueryFailure(e)),
+        e => fromEither(e).mapLeft(toQueryFailure)
+      )
+      .chain(maybeResults =>
+        maybeResults.fold(
+          // if the iterator content is none, exit the recursion
+          taskEither.of([]),
+          items =>
+            // executes backup&delete for this set of items
+            array.sequence(taskEither)(
+              items.map((item: T) =>
+                sequenceT(taskEitherSeq)<
+                  DataFailure,
+                  // tslint:disable-next-line: readonly-array
+                  [
+                    TaskEither<DataFailure, T>,
+                    TaskEither<DataFailure, string>,
+                    // tslint:disable-next-line: readonly-array
+                    TaskEither<DataFailure, readonly T[]>
+                  ]
+                >(
+                  saveDataToBlob<T>(
+                    userDataBackup,
+                    makeBackupBlobName(item),
+                    item
+                  ),
+                  tryCatch(() => deleteSingle(item), toError)
+                    .mapLeft(toDocumentDeleteFailure)
+                    .chain(_ => fromEither(_).mapLeft(toDocumentDeleteFailure)),
+                  // recursive step
+                  executeRecursiveBackupAndDelete<T>(
```

Comment:
Is this basically a reduce operation on the iteration?

Reply:
It's basically [this](https://github.com/pagopa/io-functions-commons/blob/b99194e201b015e4bce9c4dec7692afadc6d8d9a/src/utils/documentdb.ts#L544), but I didn't get to fit it in at the time of writing